From 7d257d0d36108c91527b4d626bf17fe69848bc8b Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Sat, 4 Jul 2020 12:45:50 +0200
Subject: [PATCH] pon_qos: fix cpu port queue lookup map

DPM configures the re-insert queue map internally. It requires the following
flags to be set in the lookup table enc=1,dec=0, mpe1=0, mpe2=0 as this flags
are set in the DMA descriptor for the re-inserted packets. The fix avoids
setting the flags to 1 for the cpu port. This is only done for all other ports.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c | 29 ++++++++++++++++++++-------
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.h |  4 ++++
 2 files changed, 26 insertions(+), 7 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index a8071c4323c8..3d0e45f0fd0f 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -776,10 +776,18 @@ static int pon_qos_set_qmap(struct pon_qos_q_data *qid,
 	qmap_set.map.subif = tc;
 	qmap_set.map.class = tc;
 	qmap_set.mask.flowid = 1;
-	qmap_set.mask.enc = 1;
-	qmap_set.mask.dec = 1;
-	qmap_set.mask.mpe1 = 1;
-	qmap_set.mask.mpe2 = 1;
+
+	/* Do not set the flags below for the CPU port otherwise the lookup
+	 * table entry for the re-insert queue is overwritten. The re-inserted
+	 * packets carry enc = 1,dec = 0, mpe1 = 0 and mpe = 0 in the DMA
+	 * descriptor and they have to be kept.
+	 */
+	if (!pon_qos_is_cpu_port(port)) {
+		qmap_set.mask.enc = 1;
+		qmap_set.mask.dec = 1;
+		qmap_set.mask.mpe1 = 1;
+		qmap_set.mask.mpe2 = 1;
+	}
 
 	ret = dp_queue_map_set(&qmap_set, 0);
 	if (ret == DP_FAILURE) {
@@ -900,7 +908,6 @@ int pon_qos_update_qmap(struct net_device *dev,
 		ret = pon_qos_get_subif_idx(q_tc->indev, &tc);
 		if (ret < 0)
 			return -1;
-		/* TODO: Force TC if PON DS */
 	} else {
 		tc = q_tc->tc;
 	}
@@ -909,8 +916,8 @@ int pon_qos_update_qmap(struct net_device *dev,
 		   sch->port, sch->deq_idx, qid->qid, q_tc->handle, tc,
 		   sch->ds ? "true" : "false");
 
-	/* TODO: port zero is currently the CPU port */
-	if (!sch->port) {
+	/* TODO: DPM uses port zero currently as the CPU port */
+	if (pon_qos_is_cpu_port(sch->port)) {
 		/* CPU TC mappings are based on the class id handle for now i.e.
 		 * queue with classid X:Y is mapped tc = (Y-1) + 8.
 		 * Later this can be improved by using tc-cookie.
@@ -1166,3 +1173,11 @@ int pon_qos_shaper_remove(struct pon_qos_qdisc *sch,
 
 	return 0;
 }
+
+bool pon_qos_is_cpu_port(int port)
+{
+	if (port == PON_QOS_CPU_PORT)
+		return true;
+
+	return false;
+}
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
index d033526a84f7..4bdac6400c57 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
@@ -17,6 +17,7 @@
 #define PON_QOS_SCH_MAX_LEVEL 3
 #define PON_QOS_8021P_HIGHEST_PRIO 7
 #define PON_QOS_UNUSED -1
+#define PON_QOS_CPU_PORT 0
 
 enum pon_qos_qdisc_type {
 	PON_QOS_QDISC_NONE,
@@ -184,4 +185,7 @@ int pon_qos_shaper_add(struct pon_qos_qdisc *sch,
 		       struct tc_tbf_qopt_offload_params *params);
 int pon_qos_shaper_remove(struct pon_qos_qdisc *sch,
 			  struct pon_qos_q_data *qid);
+
+bool pon_qos_is_cpu_port(int port);
+
 #endif
